juju/juju
# Add visibility and introspectability to the machine lock.

There have been many times where there have been subtle errors in juju, or the charms it deploys, and it is very hard to track down.  This branch adds some visibility to the hook execution lock, both in a log file that is written out, and in the introspection endpoint.

To make this possible I first had to add a core/machinelock package that is the central place to manage the lock.  All call sites have been updated to use this package. Both the machine agent and the unit agent create an instance and pass it in through manifold config.  The juju-run executable also creates one as needed.

This change is landing in 2.3 first as we will most likely do another 2.3 release at some stage, and this would be good to have there. I'll forward port it once landed.

## QA steps

* bootstrap
* deploy ubuntu-lite
* add two more units of ubuntu-lite to the same machine
* ssh into the machine
* look at /var/log/juju/machine-lock.log, you should see something like (although it will be different, I had to upgrade due to missing the agent name):
```
2018-07-22 01:31:46 === agent unit-ubuntu-lite-2 started ===
2018-07-22 01:31:47 unit-ubuntu-lite-2: uniter (run leader-settings-changed hook), waited 0s, held 0s
2018-07-22 01:31:46 === agent unit-ubuntu-lite-0 started ===
2018-07-22 01:31:47 unit-ubuntu-lite-0: uniter (run config-changed hook), waited 0s, held 0s
2018-07-22 01:31:47 unit-ubuntu-lite-2: uniter (run config-changed hook), waited 0s, held 0s
2018-07-22 01:31:46 === agent unit-ubuntu-lite-1 started ===
2018-07-22 01:31:48 unit-ubuntu-lite-1: uniter (run leader-settings-changed hook), waited 0s, held 0s
2018-07-22 01:31:48 unit-ubuntu-lite-1: uniter (run config-changed hook), waited 0s, held 0s
```
* run `juju-machine-lock`
```
$ juju-machine-lock 
machine-0:
  holder: none
unit-ubuntu-lite-0:
  holder: none
unit-ubuntu-lite-1:
  holder: none
unit-ubuntu-lite-2:
  holder: none
```

## Documentation changes

These are development and debugging tools only. Perhaps we should make sure we have a section for that at some time.

## Bug reference

While this isn't fixing any particular bug now, I expect it will help us debug others.
